packages/notte-core/src/notte_core/credentials/base.py

```diff
@@ -625,36 +414,108 @@ def replace_credentials(
-                action.params_values = self.replace_placeholder_credentials_in_param_values(params_values, attrs, creds)
-                return action
-            case FillAction(value=value):
-                action.value = self.replace_placeholder_credentials(value, attrs, creds)
-                return action
-            case _:
-                return action
 
-    def system_instructions(self) -> str:
-        return """CRITICAL: In FillAction, write strictly the information provided, everything has to match exactly."""
+        if not isinstance(action, (FillAction, SelectDropdownOptionAction)):
+            raise ValueError(f"Cant put credentials for action type {type(action)}")
 
-    def instructions(self) -> str:
-        """Get the credentials system prompt."""
-        website_instructs = """Sign-in & Sign-up instructions:
-If you are asked for credentials for signing in or up:"""
+        placeholder_value = get_str_value(action.value)
+        cred_class = CredentialField.placeholder_map[placeholder_value]
+        cred_key = cred_class.alias
 
-        for cred in self.get_singleton_credentials():
-            website_instructs += cred.instructions()
-            website_instructs += "\n"
+        if cred_class in (CardHolderField, CardNumberField, CardCVVField, CardFullExpirationField):
+            creds_dict = self.get_credit_card()
+        else:
+            creds_dict = self.get_credentials(snapshot.metadata.url)
 
-        for cred_type in CredentialField.registry.values():
-            if not cred_type.singleton and hasattr(cred_type, "placeholder_value"):
-                website_instructs += cred_type.default_instructions(cred_type.placeholder_value)
-                website_instructs += "\n"
+            if creds_dict is None:
+                raise ValueError(f"No credentials found in vault for url={snapshot.metadata.url}")
+
+        cred_value = creds_dict.get(cred_key)
+
+        if cred_value is None:
+            raise ValueError(f"No credential of type {cred_key} found in vault")
+
+        if not isinstance(cred_value, str):
+            raise ValueError(f"Invalid type for credential: {type(cred_value)}")
 
-        return website_instructs
+        cred_instance = cred_class(value=cred_value)
+
+        validate_element = cred_instance.validate_element(attrs)
+
+        if not validate_element:
+            logger.warning(f"Could not validate element with attrs {attrs} for {cred_key}")
+        else:
+            action.value = ValueWithPlaceholder(cred_value, placeholder_value)
+
+        return action
+
+    def system_instructions(self) -> str:
+        return """CRITICAL: In FillAction, write strictly the information provided, everything has to match exactly."""
+
+    def instructions(self) -> str:
+        return f"""CREDENTIAL HANDLING MODULE
+==========================
+
+When encountering forms that request sign-in or authentication information:
+
+EMAIL CREDENTIALS:
```

Comment:
TODO: add a test to check that each credential type defined is present is this prompt